### DISCUSSION TITLE
Memory use for large TIFF image

### DISCUSSION THREAD

**Discussion by jcupitt** [libvips/libvips#1991] [Q&A]

**Comment by Witten2026** [comment]
Hallo,
we have the following problem: Laptop with 32GB ram, Intel Core Ulta 7 225H, grafic card 128 MB, hard disk colume 954 GB. We want to resize an image with about 10 GB or 147460 X 57340 pixel. In the first and second attempt we tried with n image with 3 bands and are getting the massage: for the first (vips:23628): vips-warning ** 14:17:08:487: out of memory -- size  == 4MB for the second: (vips:5360): vips-warning ** 14:19:28:798: out of memory ....  in the last attemps we used the iamge iamge but in this time with 4 bands wich results to the message: (vips:19308): Glib-Error **: 14:21:37:078: ../glib-2.89.4/glib/gmem.c:139: failed to allocate 37748736 bytes.

Please find atteched the corresponding screenshot
<img width="1529" height="904" alt="error meldung speicher" src="https://github.com/user-attachments/assets/500accdf-df37-4ca4-85b6-59b4638cc69a" />

Kinds regards
Tilman Voigt

**Comment by jcupitt** [comment]
Hi Tilman,

It's probably an issue with the exact format the image has been written in. TIFF is extremely complicated and there are dozens of ways to encode pixels, not all of which can be read efficiently.

Could you run tiffinfo on your image file? It prints a lot of data showing the structure of the file.

http://wildtramper.com/sw/tiffinfo/tiffinfo.html

Then copy-paste the output into this discussion (don't post a screenshot!).

If you can control the format of `Pf.tiff`, then tiled TIFF with LZW compression is probably the best.

**Comment by Witten2026** [comment]
Hallo Ninfa,

hier die Antwort von J. Cupitt mit den ersten Schritten die wir gehen müssen/sollen damit wir mehr erfahren. Das probieren wir morgen mal aus.

Gruß Tilman

------ Originalnachricht ------
Von "John Cupitt" ***@***.******@***.***>>
An "libvips/libvips" ***@***.******@***.***>>
Cc "Voigt, Tilman" ***@***.******@***.***>>; "Comment" ***@***.******@***.***>>
Datum 15.09.2026 18:22:26
Betreff Re: [libvips/libvips] Memory use for large TIFF image (Discussion #1991)


ACHTUNG: EXTERNE E-Mail! Klicken Sie nicht auf Links und öffnen Sie keine Anhänge, falls Sie nicht sicher sind, dass der Absender vertrauenswürdig ist.



Hi Tilman,

It's probably an issue with the exact format the image has been written in. TIFF is extremely complicated and there are dozens of ways to encode pixels, not all of which can be read efficiently.

Could you run tiffinfo on your image file? It prints a lot of data showing the structure of the file.

http://wildtramper.com/sw/tiffinfo/tiffinfo.html

—
Reply to this email directly, view it on GitHub<https://github.com/libvips/libvips/discussions/1991?email_source=notifications&email_token=B6GXLGR2ALWTS6HIRVQVKGL5PFUEFA5CNFSNUABIM5UWIORPF5TWS5BNNB2WEL2ENFZWG5LTONUW63SDN5WW2ZLOOQXTCOBUGUZDMNBVUZZGKYLTN5XKOY3PNVWWK3TUUVSXMZLOOSWGM33PORSXEX3DNRUWG2Y#discussioncomment-18452645>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/B6GXLGXQZELYWRMJT2WTWRL5PFUEFAVCNFSNUABFKJSXA33TNF2G64TZHMYTEOJRGQYTAO2ENFZWG5LTONUW63R3GMZDENJYGAZ2C5QC>.
Triage notifications, keep track of coding agent tasks and review pull requests on the go with GitHub Mobile for iOS<https://github.com/notifications/mobile/ios/B6GXLGVLH53E5D3ER3GREDL5PFUEFA5CNFSNUABIM5UWIORPF5TWS5BNNB2WEL2ENFZWG5LTONUW63SDN5WW2ZLOOQXTCOBUGUZDMNBVUZZGKYLTN5XKOY3PNVWWK3TUUVSXMZLOOSVGM33PORSXEX3JN5ZQ> and Android<https://github.com/notifications/mobile/android/B6GXLGSRRQUXCIYQBZ7EW4T5PFUEFA5CNFSNUABIM5UWIORPF5TWS5BNNB2WEL2ENFZWG5LTONUW63SDN5WW2ZLOOQXTCOBUGUZDMNBVUZZGKYLTN5XKOY3PNVWWK3TUUVSXMZLOOSXGM33PORSXEX3BNZSHE33JMQ>. Download it today!
You are receiving this because you commented.Message ID: ***@***.***>

[https://www.uni-wh.de/fileadmin/user_upload/offsite/infotag-banner-e-mail-signatur.png] <https://www.uni-wh.de/dein-studium/infoangebote/infotag?mtm_campaign=infotag&mtm_source=email&mtm_medium=email&mtm_content=organic-infotag-interessierte-signatur>

[https://www.uni-wh.de/fileadmin/user_upload/offsite/uwh-logo-here-we-grow-blau-grau-signatur.png]

Private Universität Witten/Herdecke gGmbH
Alfred-Herrhausen-Straße 50
D - 58448 Witten

Homepage: www.uni-wh.de<https://www.uni-wh.de>

[https://www.uni-wh.de/fileadmin/user_upload/00_Allgemeines/4_Social_Media_Icons/email-signatur-some-icons_instagram.png] <https://www.instagram.com/uni_wittenherdecke>     [https://www.uni-wh.de/fileadmin/user_upload/00_Allgemeines/4_Social_Media_Icons/email-signatur-some-icons_tiktok.png]  ***@***.***>     [https://www.uni-wh.de/fileadmin/user_upload/00_Allgemeines/4_Social_Media_Icons/e-mail-signatur-some-icons_linkedin.png]  <https://de.linkedin.com/school/universit%C3%A4t-witten-herdecke/>     [https://www.uni-wh.de/fileadmin/user_upload/00_Allgemeines/4_Social_Media_Icons/email-signatur-some-icons_facebook.png]  <https://facebook.com/UniWH>     [https://www.uni-wh.de/fileadmin/user_upload/00_Allgemeines/4_Social_Media_Icons/email-signatur-some-icons_youtube.png]  <https://www.youtube.com/user/universitywitten>     [https://www.uni-wh.de/fileadmin/user_upload/00_Allgemeines/4_Social_Media_Icons/email-signatur-some-icons_blog.png]  <https://blog.uni-wh.de>

Geschäftsführung: Prof. Dr. Martin Butzlaff (Präsident), Dipl. oec. Jan Peter Nonnenkamp (Kanzler)

Sitz der Gesellschaft: Witten
Handelsregister des Amtsgerichts Bochum Nr. HRB 8671

**Comment by jcupitt** [comment]
Wait, are these microscope slides?

You can resize them directly. For example, I have this SVS:

```
$ vipsheader x.svs
x.svs: 127488x92162 uchar, 4 bands, srgb, openslideload
```

I can resize with:

```
$ vips resize x.svs[rgb] y.tif 0.1 --vips-progress --vips-leak
vips temp-19: 12749 x 9216 pixels, 32 threads, 128 x 128 tiles, 640 lines in buffer
vips temp-19: done in 306s          
vips_threadset_free: peak of 20 threads
memory: high-water mark 2.44 GB
```

(`--vips-leak` makes it track memory use and thread use)

Microscope slides are pyramids, so you can resize much more efficiently with `thumbnail`:

```
$ vips thumbnail x.svs[rgb] y.tif 12749 --vips-progress --vips-leak
vips temp-16: 12749 x 9216 pixels, 32 threads, 128 x 128 tiles, 640 lines in buffer
vips temp-16: done in 63.3s          
vips_threadset_free: peak of 19 threads
memory: high-water mark 373.94 MB
```

Now it's only 400mb and 60 seconds.

**Comment by jcupitt** [comment]
Sorry, last one, you can use `--vips-concurrency=1` for a good drop in memory use.

Normally, libvips will try to use all your threads for processing. However, resize does not need much CPU and each thread needs a lot of memory for processing. In this case, using just one thread is often better.

For example, with a tiled LZW image I see:

```
$ vips resize tiled-lzw.tif[unlimited] y.tif 0.1 --vips-progress --vips-leak
vips temp-19: 14746 x 5734 pixels, 32 threads, 128 x 128 tiles, 640 lines in buffer
vips temp-19: done in 118s          
vips_threadset_free: peak of 20 threads
memory: high-water mark 2.68 GB
```

And with one thread:

```
$ vips resize tiled-lzw.tif[unlimited] y.tif 0.1 --vips-progress --vips-leak --vips-concurrency=1
vips temp-19: 14746 x 5734 pixels, 1 threads, 128 x 128 tiles, 128 lines in buffer
vips temp-19: done in 113s          
vips_threadset_free: peak of 3 threads
memory: high-water mark 766.62 MB
```

A little faster, and 3x less memory.

**Comment by Witten2026** [comment]
Dear John,

I'm transfering you some screenshots from my Terra mobile 1470 laptop. We have tested two images. The smaller one is the result of a conversion from a file in .vmic format (PreciPoint). The conversion to bigtiff leads to the file PL.tiff. The flattening with vips produces the file Pf.tiff; the 10% resize produces zielbild.tiff. With tiffinfo we are getting the file infos - see results_with_small_immage.png.

The huge immage also converted from file format .vmic to PL1.tiff, flattened to Pf1.tiff will not open with tiffinfo. After 10% resized to zielbild1.tiff tiffinfo is showing the information - see screenshot results_with_huge_image_on_terra_laptop.png .

The screenshot results_with_huge_image.png shows the results after converting, the same image as in PL1, Pf1, zeilbid1.tiff, on an new Lenovo ThinkBook. The files are named PL2.tiff Pf2.tiff and zeilbild2.tiff and zielbild2b.tiff. File PL2.tiff and Pf2.tiff and zielbidl2.tiff are produced on the Lenovo laptop. File zielbild2b.tiff was produced on the Terra laptop. As with the PL1.tiff and Pf1.tiff files the PL2 and Pf2.tiff files are showing the unknown tiff signature. The file zielbild2.tiff is the result after the attempt to resize the PL2.tiff file which results the error messages of yesterday. Tiffinfo cannot open this file. The file zielbild2b.tiff is the result of the 10% resize of the Pf2.tiff image on the Terra laptop. This resize worked well and is showing the same tiffinfo as zielbidl1.tiff.

the orinial image xy.vmic has 4,08 GB
PL1.tiff and PL2.tiff have about 10,57 GB with 147456 x 57344 pixel
Pf1.tiff and Pf2.tiff have about 24,77 GB with 147456 x 57344 pixel

zielbild1.tiff and zielbild2b.tiff have 247 MB

zielbild2.tiff has only 15 MB

For running the tiffinfo I had to download additionally the file cc3260mt.dll

Kind regards  Tilman Voigt

------ Originalnachricht ------
Von "John Cupitt" ***@***.******@***.***>>
An "libvips/libvips" ***@***.******@***.***>>
Cc "Voigt, Tilman" ***@***.******@***.***>>; "Comment" ***@***.******@***.***>>
Datum 16.09.2026 10:33:28
Betreff Re: [libvips/libvips] Memory use for large TIFF image (Discussion #1991)


ACHTUNG: EXTERNE E-Mail! Klicken Sie nicht auf Links und öffnen Sie keine Anhänge, falls Sie nicht sicher sind, dass der Absender vertrauenswürdig ist.



Sorry, last one, you can use --vips-concurrency=1 for a good drop in memory use.

Normally, libvips will try to use all your threads for processing. However, resize does not need much CPU and each thread needs a lot of memory for processing. In this case, using just one thread is often better.

For example, with a tiled LZW image I see:

$ vips resize tiled-lzw.tif[unlimited] y.tif 0.1 --vips-progress --vips-leak
vips temp-19: 14746 x 5734 pixels, 32 threads, 128 x 128 tiles, 640 lines in buffer
vips temp-19: done in 118s
vips_threadset_free: peak of 20 threads
memory: high-water mark 2.68 GB


And with one thread:

$ vips resize tiled-lzw.tif[unlimited] y.tif 0.1 --vips-progress --vips-leak --vips-concurrency=1
vips temp-19: 14746 x 5734 pixels, 1 threads, 128 x 128 tiles, 128 lines in buffer
vips temp-19: done in 113s
vips_threadset_free: peak of 3 threads
memory: high-water mark 766.62 MB


A little faster, and 3x less memory.

—
Reply to this email directly, view it on GitHub<https://github.com/libvips/libvips/discussions/1991?email_source=notifications&email_token=B6GXLGVNFM3S4T54VO5V7735PJF5RA5CNFSNUABIM5UWIORPF5TWS5BNNB2WEL2ENFZWG5LTONUW63SDN5WW2ZLOOQXTCOBUGYYTSOBQUZZGKYLTN5XKOY3PNVWWK3TUUVSXMZLOOSWGM33PORSXEX3DNRUWG2Y#discussioncomment-18461980>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/B6GXLGX32UYBI3BOSCBXMH35PJF5RAVCNFSNUABFKJSXA33TNF2G64TZHMYTEOJRGQYTAO2ENFZWG5LTONUW63R3GMZDENJYGAZ2C5QC>.
Triage notifications, keep track of coding agent tasks and review pull requests on the go with GitHub Mobile for iOS<https://github.com/notifications/mobile/ios/B6GXLGSPW2LAMFFS5RSBTV35PJF5RA5CNFSNUABIM5UWIORPF5TWS5BNNB2WEL2ENFZWG5LTONUW63SDN5WW2ZLOOQXTCOBUGYYTSOBQUZZGKYLTN5XKOY3PNVWWK3TUUVSXMZLOOSVGM33PORSXEX3JN5ZQ> and Android<https://github.com/notifications/mobile/android/B6GXLGVHL7JARHQRUGMHXML5PJF5RA5CNFSNUABIM5UWIORPF5TWS5BNNB2WEL2ENFZWG5LTONUW63SDN5WW2ZLOOQXTCOBUGYYTSOBQUZZGKYLTN5XKOY3PNVWWK3TUUVSXMZLOOSXGM33PORSXEX3BNZSHE33JMQ>. Download it today!
You are receiving this because you commented.Message ID: ***@***.***>

[https://www.uni-wh.de/fileadmin/user_upload/offsite/infotag-banner-e-mail-signatur.png] <https://www.uni-wh.de/dein-studium/infoangebote/infotag?mtm_campaign=infotag&mtm_source=email&mtm_medium=email&mtm_content=organic-infotag-interessierte-signatur>

[https://www.uni-wh.de/fileadmin/user_upload/offsite/uwh-logo-here-we-grow-blau-grau-signatur.png]

Private Universität Witten/Herdecke gGmbH
Alfred-Herrhausen-Straße 50
D - 58448 Witten

Homepage: www.uni-wh.de<https://www.uni-wh.de>

[https://www.uni-wh.de/fileadmin/user_upload/00_Allgemeines/4_Social_Media_Icons/email-signatur-some-icons_instagram.png] <https://www.instagram.com/uni_wittenherdecke>     [https://www.uni-wh.de/fileadmin/user_upload/00_Allgemeines/4_Social_Media_Icons/email-signatur-some-icons_tiktok.png]  ***@***.***>     [https://www.uni-wh.de/fileadmin/user_upload/00_Allgemeines/4_Social_Media_Icons/e-mail-signatur-some-icons_linkedin.png]  <https://de.linkedin.com/school/universit%C3%A4t-witten-herdecke/>     [https://www.uni-wh.de/fileadmin/user_upload/00_Allgemeines/4_Social_Media_Icons/email-signatur-some-icons_facebook.png]  <https://facebook.com/UniWH>     [https://www.uni-wh.de/fileadmin/user_upload/00_Allgemeines/4_Social_Media_Icons/email-signatur-some-icons_youtube.png]  <https://www.youtube.com/user/universitywitten>     [https://www.uni-wh.de/fileadmin/user_upload/00_Allgemeines/4_Social_Media_Icons/email-signatur-some-icons_blog.png]  <https://blog.uni-wh.de>

Geschäftsführung: Prof. Dr. Martin Butzlaff (Präsident), Dipl. oec. Jan Peter Nonnenkamp (Kanzler)

Sitz der Gesellschaft: Witten
Handelsregister des Amtsgerichts Bochum Nr. HRB 8671

**Reply by jcupitt** [reply]
Please don't post screenshots!

I can't search them, and I can't copy-paste commands out of them to try myself. They are a bit useless.

Instead, select the text in your terminal and copy-paste the text into this github issue. You can put three backticks around the text like this:

```
    ```
    some text
    ```
```

and it displays like this:

```
some text
```

**Comment by Witten2026** [comment]
Sorry about that.

The following messages corresponds to the description in the last email.

1) Messages for the small image

```
Converted file from xy.vmic to PL.TIFF
C:\Users\Til\Downloads\tiffinfo>tiffinfo PL.tiff
Unknown Tiff signature in file <PL.tiff>

Flattend image from PL.tiff to Pf.tiff
C:\Users\Til\Downloads\tiffinfo>tiffinfo Pf.tiff
Pf.tiff: FileID=II, Ver=2A, IFD #1 w/ 15 Tags
 TagName(ID,Type,Len,Ofst)          =Value
 ImageWidth(0100,3,1,2CA00012)      =30464
 ImageLength(0101,3,1,2CA0001E)     =8192
 BitsPerSample(0102,3,3,2CA000D2)   =8 8 8
 Compression(0103,3,1,2CA00036)     =1  {NoCompression}
 Photometric(0106,3,1,2CA00042)     =2  {RGB}
 StripOffsets(0111,4,40,2CA001D8)   =8 11698184 23396360 35094536 ...
 Orientation(0112,3,1,2CA0005A)     =1  {TopLeft, Normal}
 SamplesPerPixel(0115,3,1,2CA00066) =3
 RowsPerStrip(0116,3,1,2CA00072)    =128
 StripByteCounts(0117,4,40,2CA000D8)=11698176 11698176 11698176 11698176 ...
 XResolution(011A,5,1,2CA000C2)     =8795517/64 (=137430)
 YResolution(011B,5,1,2CA000CA)     =8795517/64 (=137430)
 PlanarConfig(011C,3,1,2CA000A2)    =1  {ChunkyFormat}
 ResolutionUnit(0128,3,1,2CA000AE)  =2  {InchUnits}
 SampleFormat(0153,3,3,2CA002D8)    =1 1 1

10% Resized image from Pf.tiff to zielbild.tiff
C:\Users\Til\Downloads\tiffinfo>tiffinfo zielbild.tiff
zielbild.tiff: FileID=II, Ver=2A, IFD #1 w/ 15 Tags
 TagName(ID,Type,Len,Ofst)       =Value
 ImageWidth(0100,3,1,723288)     =3046
 ImageLength(0101,3,1,723294)    =819
 BitsPerSample(0102,3,3,723348)  =8 8 8
 Compression(0103,3,1,7232AC)    =1  {NoCompression}
 Photometric(0106,3,1,7232B8)    =2  {RGB}
 StripOffsets(0111,4,7,72336A)   =8 1169672 2339336 3509000 4678664 ...
 Orientation(0112,3,1,7232D0)    =1  {TopLeft, Normal}
 SamplesPerPixel(0115,3,1,7232DC)=3
 RowsPerStrip(0116,3,1,7232E8)   =128
 StripByteCounts(0117,4,7,72334E)=1169664 1169664 1169664 1169664 1169664 ...
 XResolution(011A,5,1,723338)    =8795517/64 (=137430)
 YResolution(011B,5,1,723340)    =8795517/64 (=137430)
 PlanarConfig(011C,3,1,723318)   =1  {ChunkyFormat}
 ResolutionUnit(0128,3,1,723324) =2  {InchUnits}
 SampleFormat(0153,3,3,723386)   =1 1 1
```

2) Messages for the huge image; conversion and resizing done on Terra Mobiel 1470 Laptop with 40GB RAM DDR4

```
Converted image from xy.vmic to PL1.tiff (PL1.tiff about 10 GB with about 147460 x 57340 pixel; Pf1.tiff image about 24,7GB)
C:\Users\Til\Downloads\tiffinfo>tiffinfo PL1.tiff
Unknown Tiff signature in file <PL1.tiff>

Flattened image from PL1.tiff to Pf1.tiff
C:\Users\Til\Downloads\tiffinfo>tiffinfo Pf1.tiff
Unknown Tiff signature in file <Pf1.tiff>

10% resized iamge form Pf1.tiff to zielbild1.tigg
C:\Users\Til\Downloads\tiffinfo>tiffinfo zielbild1.tiff
zielbild1.tiff: FileID=II, Ver=2A, IFD #1 w/ 15 Tags
 TagName(ID,Type,Len,Ofst)         =Value
 ImageWidth(0100,3,1,F1E8E26)      =14746
 ImageLength(0101,3,1,F1E8E32)     =5734
 BitsPerSample(0102,3,3,F1E8EE6)   =8 8 8
 Compression(0103,3,1,F1E8E4A)     =1  {NoCompression}
 Photometric(0106,3,1,F1E8E56)     =2  {RGB}
 StripOffsets(0111,4,2D,F1E8FA0)   =8 5662472 11324936 16987400 22649864 ...
 Orientation(0112,3,1,F1E8E6E)     =1  {TopLeft, Normal}
 SamplesPerPixel(0115,3,1,F1E8E7A) =3
 RowsPerStrip(0116,3,1,F1E8E86)    =128
 StripByteCounts(0117,4,2D,F1E8EEC)=5662464 5662464 5662464 5662464 ...
 XResolution(011A,5,1,F1E8ED6)     =8795517/64 (=137430)
 YResolution(011B,5,1,F1E8EDE)     =8795517/64 (=137430)
 PlanarConfig(011C,3,1,F1E8EB6)    =1  {ChunkyFormat}
 ResolutionUnit(0128,3,1,F1E8EC2)  =2  {InchUnits}
 SampleFormat(0153,3,3,F1E9054)    =1 1 1
```

3) Messages for the huge image - same image as in 2 (PL2.tiff about 10 GB with about 147460 x 57340 pixel; Pf2.tiff image about 24,7GB); conversion done on Lenovo Laptop with 32GB RAM DDR5
File after conversion

```
C:\Users\Til\Downloads\tiffinfo>tiffinfo  PL2.tiff
Unknown Tiff signature in file <PL2.tiff>

File after flattening
C:\Users\Til\Downloads\tiffinfo>tiffinfo  Pf2.tiff
Unknown Tiff signature in file <Pf2.tiff>

File after 10% resizing on the Lenovo Laptop with error massages (sent you yesterday)
C:\Users\Til\Downloads\tiffinfo>tiffinfo  zielbild2.tiff

File after 10% resizing on the Terra Mobile Laptop
C:\Users\Til\Downloads\tiffinfo>tiffinfo  zielbild2b.tiff
zielbild2b.tiff: FileID=II, Ver=2A, IFD #1 w/ 15 Tags
 TagName(ID,Type,Len,Ofst)         =Value
 ImageWidth(0100,3,1,F1E8E26)      =14746
 ImageLength(0101,3,1,F1E8E32)     =5734
 BitsPerSample(0102,3,3,F1E8EE6)   =8 8 8
 Compression(0103,3,1,F1E8E4A)     =1  {NoCompression}
 Photometric(0106,3,1,F1E8E56)     =2  {RGB}
 StripOffsets(0111,4,2D,F1E8FA0)   =8 5662472 11324936 16987400 22649864 ...
 Orientation(0112,3,1,F1E8E6E)     =1  {TopLeft, Normal}
 SamplesPerPixel(0115,3,1,F1E8E7A) =3
 RowsPerStrip(0116,3,1,F1E8E86)    =128
 StripByteCounts(0117,4,2D,F1E8EEC)=5662464 5662464 5662464 5662464 ...
 XResolution(011A,5,1,F1E8ED6)     =8795517/64 (=137430)
 YResolution(011B,5,1,F1E8EDE)     =8795517/64 (=137430)
 PlanarConfig(011C,3,1,F1E8EB6)    =1  {ChunkyFormat}
 ResolutionUnit(0128,3,1,F1E8EC2)  =2  {InchUnits}
 SampleFormat(0153,3,3,F1E9054)    =1 1 1
```

Kind regards
                 Tilman Voigt
